pkg/server/internal/resultset/cursor.go

```diff
@@ -53,3 +97,81 @@ type FetchNotifier interface {
 	// it will be used in server-side cursor.
 	OnFetchReturned()
 }
+
+var _ CursorResultSet = &tidbLazyCursorResultSet{}
+
+type tidbLazyCursorResultSet struct {
+	ResultSet
+	iter lazyRowIterator
+}
+
+// WrapWithLazyCursor wraps a ResultSet into a CursorResultSet
+func WrapWithLazyCursor(rs ResultSet, capacity, maxChunkSize int) CursorResultSet {
+	chk := chunk.New(rs.FieldTypes(), capacity, maxChunkSize)
+
+	return &tidbLazyCursorResultSet{
+		ResultSet: rs,
+		iter: lazyRowIterator{
+			rs:  rs,
+			chk: chk,
+		},
+	}
+}
+
+func (tcrs *tidbLazyCursorResultSet) GetRowIterator() RowIterator {
+	return &tcrs.iter
+}
+
+type lazyRowIterator struct {
+	rs       ResultSet
+	err      error
+	chk      *chunk.Chunk
+	idxInChk int
+	started  bool
+}
+
+func (iter *lazyRowIterator) Next(ctx context.Context) chunk.Row {
+	iter.idxInChk++
```

Comment:
If we call `Next` before `Current`, seems `Current` will returns a different row with the row returned by `Next`.

Should we init the first row when we find `!started` here?

Another way to implement this iterator is to init the first when `lazyRowIterator` is created so that we do not need to add a extra ctx argument to `Current`

Reply:
I'll add a `iter.started = true` in the `.Next()` function. Then it'll fix the difference between the first `Next` and the following `Current`. Their results are different because the second `Current()` will call an extra `Next()` because the `iter.started` is not set to `true`.

> Should we init the first row when we find `!started` here?

In `.Next()`, it'll always init the first row, because initially the chunk is empty and `if iter.idxInChk >= iter.chk.NumRows() {` always meets.

> Another way to implement this iterator is to init the first when lazyRowIterator is created so that we do not need to add a extra ctx argument to Current

Then the first `.Next()` of the executor will be called in the `EXECUTE` but not `FETCH` :thinking:. It's OK, but I don't think remove one argument of the `Current` is valuable enough to bring this change.